filters/go_impl.go

```diff
@@ -0,0 +1,416 @@
+		target, err := asString(target0)
+		if err != nil {
+			return err
+		}
+		pattern, err := asString(pattern0)
+		if err != nil {
+			return err
+		}
+		return boolOrErr(f(strings.ToLower(target), strings.ToLower(pattern)))
+	}
+}
+
+func wrapS2LCI(f func(target string, patterns []string) (bool, error)) functions.BinaryOp {
+	return func(target0, patterns0 ref.Val) ref.Val {
+		target, err := asString(target0)
+		if err != nil {
+			return err
+		}
+		patterns, err := asStringList(patterns0)
+		if err != nil {
+			return err
+		}
+		return boolOrErr(f(strings.ToLower(target), lowerList(patterns)))
+	}
+}
+
+func wrapL2SCI(f func(targets []string, pattern string) (bool, error)) functions.BinaryOp {
+	return func(targets0, pattern0 ref.Val) ref.Val {
+		targets, err := asStringList(targets0)
+		if err != nil {
+			return err
+		}
+		pattern, err := asString(pattern0)
+		if err != nil {
+			return err
+		}
+		return boolOrErr(f(lowerList(targets), strings.ToLower(pattern)))
+	}
+}
+
+func wrapL2LCI(f func(targets, patterns []string) (bool, error)) functions.BinaryOp {
+	return func(targets0, patterns0 ref.Val) ref.Val {
+		targets, err := asStringList(targets0)
+		if err != nil {
+			return err
+		}
+		patterns, err := asStringList(patterns0)
+		if err != nil {
+			return err
+		}
+		return boolOrErr(f(lowerList(targets), lowerList(patterns)))
+	}
+}
+
+func goExistsEqualsStringToString(target, pattern string) (bool, error) {
+	return target == pattern, nil
+}
+
+func goExistsEqualsStringToList(target string, patterns []string) (bool, error) {
+	for _, pattern := range patterns {
+		if target == pattern {
+			return true, nil
+		}
+	}
+	return false, nil
+}
+
+func goExistsEqualsListToString(targets []string, pattern string) (bool, error) {
+	for _, target := range targets {
+		if target == pattern {
+			return true, nil
+		}
+	}
+	return false, nil
+}
+
+func makeSet(items []string) map[string]struct{} {
+	set := make(map[string]struct{}, len(items))
+	for _, item := range items {
+		set[item] = struct{}{}
+	}
+	return set
+}
+
+func goExistsEqualsListToList(targets []string, patterns []string) (bool, error) {
+	set := makeSet(patterns)
+	for _, target := range targets {
+		if _, has := set[target]; has {
+			return true, nil
+		}
+	}
+	return false, nil
+}
+
+func goExistsStartsStringToString(target, pattern string) (bool, error) {
+	return strings.HasPrefix(target, pattern), nil
+}
+
+func goExistsStartsStringToList(target string, patterns []string) (bool, error) {
+	for _, pattern := range patterns {
+		if strings.HasPrefix(target, pattern) {
+			return true, nil
+		}
+	}
+	return false, nil
+}
+
+func goExistsStartsListToString(targets []string, pattern string) (bool, error) {
+	for _, target := range targets {
+		if strings.HasPrefix(target, pattern) {
+			return true, nil
+		}
+	}
+	return false, nil
+}
+
+func goExistsStartsListToList(targets []string, patterns []string) (bool, error) {
+	for _, pattern := range patterns {
```

Comment:
if it were me I'd probably have these `ListToList`, `ListToString`, `StringToList`, and `StringToString` handling in the wrap functions and have just `contains`, `containsCI`, `startsWith`, etc. being specific. This would make some optimizations not possible but I'm not sure how much do we need that in non-sql code

I think this is optional though and the higher priority is to get this into the codebase so I'll approve the review anyway

Reply:
We can have the functions only ListToList, so all optimizations caused by reusing some intermediate objects (like compiled RE) are possible. And derive all other methods from it